packages/xeus/src/worker.ts

```diff
@@ -173,25 +193,239 @@ export abstract class EmpackedXeusRemoteKernel extends XeusRemoteKernelBase {
+  } {
+    // copy data before making manupulations
+    let newSpecs = { ...this._currentSpecs };
+
+    /* we need to collect requested pip specs to know when we need to delete all packages
+       if a user tried to install and after they decided to delete pip packages
+       and if specs array is empty and pip specs array is not, then we should not clean an enviroment
+    */
+    let newPipSpecs = { ...this._currentPipSpecs };
+    let updatedInstalled: ISolvedPackages = { ...this._installedPackages };
+    const newInstalledPackagesMap: { [name: string]: ISolvedPackage } =
+      this.getInstalledPackagesMap();
+
+    let status: number = 1;
+
+    switch (type) {
+      case 'install':
+        newSpecs = this.addSpecs(specs, newSpecs);
+        if (!newInstalledPackagesMap['pip'] && pipSpecs.length) {
+          this.logger.error('Cannot run "pip install": pip is not installed');
+          status = 0;
+          throw new Error('Cannot run "pip install": pip is not installed');
+        } else if (pipSpecs.length) {
+          newPipSpecs = this.addSpecs(pipSpecs, newPipSpecs);
+        }
+        break;
+      case 'uninstall': {
+        const tmpSpecs: string[] = specs.length
+          ? specs
+          : pipSpecs.length
+            ? pipSpecs
+            : [];
+        let preparedPkgs: any = [];
+        const tmpNewSpecs = specs.length ? newSpecs : newPipSpecs;
+        const pkgType = specs.length ? 'conda' : 'pip';
+
+        preparedPkgs = this.canDelete(
+          tmpSpecs,
+          newInstalledPackagesMap,
+          pkgType
+        );
+
+        const specsForDeleting: string[] = [];
+        preparedPkgs.forEach((pkgData: any) => {
+          if (pkgData.status) {
+            specsForDeleting.push(pkgData.spec);
+          } else {
+            this.logger.error(pkgData.message);
+          }
+        });
+
+        if (specsForDeleting.length) {
+          const data = this.deleteSpecs(
+            specsForDeleting,
+            tmpNewSpecs,
+            newInstalledPackagesMap,
+            updatedInstalled,
+            pkgType
+          );
+          if (pkgType === 'conda') {
+            newSpecs = data.specs;
+          } else {
+            newPipSpecs = data.specs;
+          }
+          updatedInstalled = data.installed;
+        } else {
+          status = 0;
+        }
+
+        break;
+      }
+      default:
+        break;
+    }
+
+    return {
+      specs: newSpecs,
+      pipSpecs: newPipSpecs,
+      installed: updatedInstalled,
+      status
+    };
+  }
+
+  protected async solveEnv(
+    specs: string[],
+    pipSpecs: string[],
+    channels: string[],
+    installedPackages: ISolvedPackages
+  ) {
+    const newPackages = await solve({
+      ymlOrSpecs: specs,
+      installedPackages,
+      pipSpecs,
+      channels,
+      logger: this.logger
+    });
+    await this._reloadPackagesInFS({
+      ...newPackages.condaPackages,
+      ...newPackages.pipPackages
+    });
+  }
```

Comment:
Let's not wrap the `solve` function into a method, we're not gaining much.

Actually we're loosing the ability to pass options with granularity (see how `solve(options: IOptions)` vs. `solveEnv(specs: string[], pipSpecs: string[], ...))`).

So one cannot do `solve({pipSpecs: ['ipympl']})` if they only want to do a pip install run.